File: go/binary-search-tree/binary_search_tree.go
package binarysearchtree

// BinarySearchTree is a search tree.
type BinarySearchTree struct {
	data  int
	left  *BinarySearchTree
	right *BinarySearchTree
}

// Bst returns a new tree.
func Bst(d int) *BinarySearchTree {
	return &BinarySearchTree{data: d}
}

// NewBst returns a new BinarySearchTree.
func NewBst(data int) *BinarySearchTree {
	return &BinarySearchTree{data: data}
}

// Insert an element into the tree.
func (s *BinarySearchTree) Insert(d int) {
	if d <= s.data {
		if s.left != nil {
			s.left.Insert(d)
		} else {
			s.left = Bst(d)
		}
	} else {
		if s.right != nil {
			s.right.Insert(d)
		} else {
			s.right = Bst(d)
		}
	}
}

// MapString applies a string mapper to the tree.
func (s *BinarySearchTree) MapString(f func(int) string) []string {
	var r []string
	if s.left != nil {
		r = s.left.MapString(f)
	}
	r = append(r, f(s.data))
	if s.right != nil {
		for _, m := range s.right.MapString(f) {
			r = append(r, m)
		}
	}
	return r
}

// MapInt applies an int mapper to the tree.
func (s *BinarySearchTree) MapInt(f func(int) int) []int {
	var r []int
	if s.left != nil {
		r = s.left.MapInt(f)
	}
	r = append(r, f(s.data))
	if s.right != nil {
		for _, m := range s.right.MapInt(f) {
			r = append(r, m)
		}
	}
	return r
}

// SortedData returns tree data in a sorted array.
func (s *BinarySearchTree) SortedData() []int {
	return s.MapInt(func(d int) int { return d })
}

File: python/binary-search-tree/binary_search_tree.py
"""Binary Tree."""

from __future__ import annotations
from typing import cast, Generator, Generic, Iterable, Optional, TypeVar


T = TypeVar("T")


class TreeNode(Generic[T]):
    """Binary Tree Node."""

    def __init__(
        self,
        data: Optional[T],
        left: Optional[TreeNode] = None,
        right: Optional[TreeNode] = None
    ):
        self.data = data
        self.left = left
        self.right = right

    def __repr__(self) -> str:
        """Return a string representation."""
        return f"TreeNode(data={self.data}, left={self.left}, right={self.right})"

    def insert(self, value: T) -> None:
        """Insert a value into the tree."""
        if self.data is None:
            self.data = value
        elif value > self.data:  # type: ignore
            if self.right is None:
                self.right = TreeNode(None)
            self.right.insert(value)
        else:
            if self.left is None:
                self.left = TreeNode(None)
            self.left.insert(value)

    def __iter__(self) -> Generator[T, None, None]:
        """Return node values, sorted."""
        yield from self.left or []
        # if self.data is None: return
        # or this can be checked once in sorted_data() and assumed to exist here.
        yield cast(T, self.data)
        yield from self.right or []


class BinarySearchTree(Generic[T]):
    """Binary Tree."""

    def __init__(self, tree_data: Iterable[T]):
        """Initialize."""
        self.root: TreeNode[T] = TreeNode(None)
        for value in tree_data:
            self.root.insert(value)

    def data(self) -> TreeNode[T]:
        """Return Tree data."""
        return self.root

    def sorted_data(self) -> list[T]:
        """Return sorted data."""
        if self.root.data is None:
            return []
        return list(self.root)
